python/lsst/daf/butler/core/logging.py

```diff
@@ -94,3 +105,270 @@ def MDCRemove(cls, key: str) -> None:
+        """Read records from file.
+
+        Parameters
+        ----------
+        filename : `str`
+            Name of file containing the JSON records.
+
+        Notes
+        -----
+        Works with one-record-per-line format JSON files and a direct
+        serialization of the Pydantic model.
+        """
+        with open(filename, "r") as fd:
+            return cls.from_stream(fd)
+
+    @classmethod
+    def from_stream(cls, stream: IO) -> ButlerLogRecords:
+        """Read records from I/O stream.
+
+        Parameters
+        ----------
+        stream : `typing.IO`
+            Stream from which to read JSON records.
+
+        Notes
+        -----
+        Works with one-record-per-line format JSON files and a direct
+        serialization of the Pydantic model.
+        """
+        first_line = stream.readline()
+
+        if not first_line:
+            # Empty file, return zero records.
+            return cls.from_records([])
+
+        # Allow byte or str streams since pydantic supports either.
+        first_char = str(first_line[0])
+        if first_char == "[":
+            # This is a ButlerLogRecords model serialization.
+            all = first_line + stream.read()
+            return cls.parse_raw(all)
+
+        # A stream of records with one record per line.
+        if first_char != "{":
+            raise RuntimeError(f"Unrecognized JSON log format. First lines is '{first_line}'")
+        records = [ButlerLogRecord.parse_raw(first_line)]
+        for line in stream:
+            if line:  # Filter out blank lines.
+                records.append(ButlerLogRecord.parse_raw(line))
+
+        return cls.from_records(records)
+
+    @property
+    def log_format(self) -> str:
+        if self._log_format is None:
+            return _LONG_LOG_FORMAT
+        return self._log_format
+
+    @log_format.setter
+    def log_format(self, format: str) -> None:
+        self._log_format = format
+
+    def __len__(self) -> int:
+        return len(self.__root__)
+
+    # The signature does not match the one in BaseModel but that is okay
+    # if __root__ is being used.
+    # See https://pydantic-docs.helpmanual.io/usage/models/#custom-root-types
+    def __iter__(self) -> Iterator[ButlerLogRecord]:  # type: ignore
+        return iter(self.__root__)
+
+    def __setitem__(self, index: int, value: Record) -> None:
+        self.__root__[index] = self._validate_record(value)
+
+    def __getitem__(self, index: Union[slice, int]) -> Union[ButlerLogRecords, ButlerLogRecord]:
+        # Handles slices and returns a new collection in that
+        # case.
+        item = self.__root__[index]
+        if isinstance(item, list):
+            return type(self)(__root__=item)
+        else:
+            return item
+
+    def __reversed__(self) -> Iterator[ButlerLogRecord]:
+        return self.__root__.__reversed__()
+
+    def __delitem__(self, index: int) -> None:
```

Comment:
Can `index` be a slice here too?

Reply:
Maybe. I hadn't thought about it. Will look.